File: solutions/python/affine-cipher/2/affine_cipher.py
import math

def encode(plain_text: str, a: int, b: int) -> str:
    # Check if a and m are coprime
    m = 26
    if math.gcd(a, m) != 1:
        raise ValueError("a and m must be coprime.")
    
    # Process the plain text: remove non-alphanumeric characters and convert to lowercase
    processed_text = ''.join(c.lower() for c in plain_text if c.isalnum())
    
    # Encrypt each character
    encrypted_chars = []
    for char in processed_text:
        if char.isdigit():
            encrypted_chars.append(char)
        else:
            x = ord(char) - ord('a')
            y = (a * x + b) % m
            encrypted_char = chr(y + ord('a'))
            encrypted_chars.append(encrypted_char)
    
    # Group the encrypted text into chunks of 5 characters
    encrypted_text = ' '.join(
        ''.join(encrypted_chars[i:i+5]) for i in range(0, len(encrypted_chars), 5)
    )
    
    return encrypted_text
def decode(ciphered_text: str, a: int, b: int) -> str:
    # Check if a and m are coprime
    m = 26
    if math.gcd(a, m) != 1:
        raise ValueError("a and m must be coprime.")
    
    # Process the ciphered text: remove spaces and convert to lowercase
    processed_text = ''.join(c.lower() for c in ciphered_text if c.isalnum())
    
    # Find the modular multiplicative inverse of a
    a_inv = pow(a, -1, m)
    
    # Decrypt each character
    decrypted_chars = []
    for char in processed_text:
        if char.isdigit():
            decrypted_chars.append(char)
        else:
            y = ord(char) - ord('a')
            x = (a_inv * (y - b)) % m
            decrypted_char = chr(x + ord('a'))
            decrypted_chars.append(decrypted_char)
    
    # Join the decrypted characters into a single string
    decrypted_text = ''.join(decrypted_chars)
    
    return decrypted_text

File: solutions/ruby/affine-cipher/2/affine_cipher.rb
class Affine
  def initialize(a, b)
    @a = a
    @b = b
    raise ArgumentError, 'a and m must be coprime' unless coprime?(a, 26)
  end

  def encode(plaintext)
    plaintext.downcase.gsub(/[^a-z0-9]/, '').chars.map { |c| encode_char(c) }.join.scan(/.{1,5}/).join(' ')
  end

  def decode(ciphertext)
    ciphertext.gsub(/\s+/, '').chars.map { |c| decode_char(c) }.join
  end

  private

  def coprime?(a, m)
    a.gcd(m) == 1
  end

  def encode_char(c)
    if c =~ /[0-9]/ 
      c
    else
      ((@a * (c.ord - 'a'.ord) + @b) % 26 + 'a'.ord).chr
    end
  end

  def decode_char(c)
    if c =~ /[0-9]/ 
      c
    else
      ((modular_inverse(@a, 26) * (c.ord - 'a'.ord - @b)) % 26 + 'a'.ord).chr
    end
  end

  def modular_inverse(a, m)
    a, m = a % m, m
    (1..m).each do |x|
      return x if (a * x) % m == 1
    end
    raise 'Modular inverse does not exist'
  end
end